codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/XmlShapeDeserVisitor.java

```diff
@@ -0,0 +1,403 @@
+ */
+public class XmlShapeDeserVisitor extends DocumentShapeDeserVisitor {
+    private static final TimestampFormatTrait.Format DEFAULT_TIMESTAMP_FORMAT = TimestampFormatTrait.Format.DATE_TIME;
+    private static final Logger LOGGER = Logger.getLogger(XmlShapeDeserVisitor.class.getName());
+
+    private final Predicate<MemberShape> memberFilter;
+
+    /**
+     * @param context The generation context.
+     */
+    public XmlShapeDeserVisitor(GenerationContext context) {
+        this(context, FunctionalUtils.alwaysTrue());
+    }
+
+    /**
+     * @param context The generation context.
+     * @param memberFilter A filter that is applied to structure members. This is useful for
+     *     members that won't be in the body.
+     */
+    public XmlShapeDeserVisitor(GenerationContext context, Predicate<MemberShape> memberFilter) {
+        super(context);
+        this.memberFilter = memberFilter;
+    }
+
+    @Override
+    protected Map<String, String> getAdditionalArguments() {
+        return Collections.singletonMap("decoder", "*smithydecoding.XMLNodeDecoder");
+    }
+
+    private XmlMemberDeserVisitor getMemberDeserVisitor(MemberShape member, String dataDest, Predicate<Shape> filter) {
+        // Get the timestamp format to be used, defaulting to epoch seconds.
+        TimestampFormatTrait.Format format = member.getMemberTrait(getContext().getModel(), TimestampFormatTrait.class)
+                .map(TimestampFormatTrait::getFormat).orElse(DEFAULT_TIMESTAMP_FORMAT);
+
+        // Check if FlattenedTrait  is applied
+        Boolean isFlattened = member.hasTrait(XmlFlattenedTrait.ID);
+
+        return new XmlMemberDeserVisitor(getContext(), dataDest, format, filter, isFlattened);
+    }
+
+    @Override
+    protected void deserializeCollection(GenerationContext context, CollectionShape shape) {
+        GoWriter writer = context.getWriter();
+        SymbolProvider symbolProvider = context.getSymbolProvider();
+        Symbol symbol = symbolProvider.toSymbol(shape);
+
+        // Initialize the value now that the start stub has verified that there's something there.
+        writer.write("var sv $P", symbol);
+        writer.openBlock("if *v == nil {", "", () -> {
+            writer.write("sv = make($P, 0)", symbol);
+            writer.openBlock("} else {", "}", () -> {
+                writer.write("sv = *v");
+            });
+        });
+
+        // Iterate through the decoder. The member visitor will handle popping tokens.
+        writer.openBlock("for {", "}", () -> {
+            writer.write("t, done, err := decoder.Token()");
+            writer.write("if err != nil { return err }");
+            writer.write("if done { break }");
+
+            MemberShape member = shape.getMember();
+            Shape target = context.getModel().expectShape(member.getTarget());
+            String serializedMemberName = getSerializedMemberName(member);
+            writer.openBlock("switch t.Name.Local {", "}", () -> {
```

Comment:
This switch appears to only have one case and a default, so it could probably just be an if

Reply:
Just kept it switch to be consistent and not have to change the code flow specifically for this case. Switches will have similar performance to if-else in Go if not more.

Reply:
This does seem to be a pattern though - where else besides the error dispatcher do you need the switch?

Reply:
updated